mitmproxy/proxy2/layers/http2.py

```diff
@@ -0,0 +1,418 @@
+        self.client_conn = connection.H2Connection(config=client_config)
+        self.client_conn.initiate_connection()
+        yield commands.SendData(self.context.client, self.client_conn.data_to_send())
+
+        server_config = h2.config.H2Configuration(
+            client_side=True,
+            header_encoding=False,
+            validate_outbound_headers=False,
+            validate_inbound_headers=False)
+        self.server_conn = connection.H2Connection(config=server_config)
+        self.server_conn.initiate_connection()
+        yield commands.SendData(self.context.server, self.server_conn.data_to_send())
+
+        self.streams: Dict[int, Http2Stream] = dict()
+        self.server_to_client_stream_ids: Dict[int, int] = dict([(0, 0)])
+        self.unfinished_bodies: Dict[connection.H2H2Connection, tuple] = dict()
+
+        yield commands.Log("HTTP/2 connection started")
+
+        self._handle_event = self.process_data
+
+    _handle_event = start
+
+    @expect(events.DataReceived, events.ConnectionClosed)
+    def process_data(self, event: events.Event) -> commands.TCommandGenerator:
+        if isinstance(event, events.DataReceived):
+            dead = [stream for stream in self.streams.values() if stream.death_time]
+            for stream in dead:
+                if stream.death_time <= time.time() - 10:
+                    self.streams.pop(stream.stream_id, None)
+
+
+            from_client = event.connection == self.context.client
+            if from_client:
+                source = self.client_conn
+                other = self.server_conn
+                send_to_source = self.context.client
+                send_to_other = self.context.server
+            else:
+                source = self.server_conn
+                other = self.client_conn
+                send_to_source = self.context.server
+                send_to_other = self.context.client
+
+            received_h2_events = source.receive_data(event.data)
+            yield commands.SendData(send_to_source, source.data_to_send())
+
+            for h2_event in received_h2_events:
+                yield commands.Log(
+                    "HTTP/2 event from {}: {}".format("client" if from_client else "server", h2_event)
+                )
+
+                eid = None
+                if hasattr(h2_event, 'stream_id'):
+                    if not from_client and h2_event.stream_id % 2 == 1:
+                        eid = self.server_to_client_stream_ids[h2_event.stream_id]
+                    else:
+                        eid = h2_event.stream_id
+
+                if isinstance(h2_event, h2events.RequestReceived):
+                    self.streams[eid] = Http2Stream(h2_event, self.context.client, self.context.server)
+                    self.streams[eid].timestamp_start = time.time()
+                    self.streams[eid].request_arrived.set()
+
+                    yield commands.Hook("requestheaders", self.streams[eid].flow)
```

Comment:
This still has the problem of blocking the entire h2 connection, you are aware of that?